tensorboard/components/tf_tensorboard/test/tensorboardTests.ts

```
@@ -136,32 +128,60 @@ namespace tf_tensorboard {
+            origin: 'https://tensorboard.is',
+          });
+
+          tensorboard = fixture('tensorboardFixture');
           tensorboard.demoDir = 'data';
+          tensorboard.brand = 'Custom Brand';
           tensorboard.autoReloadEnabled = false;
+          tensorboard.homePath = '/awesome';
+
+          // Branding and other components use `dom-if` which updates the dom in an
+          // animation frame. Flush and remove the asynchronicity.
+          flush(done);
         });
 
-        it('uses customized brand for title', (done) => {
-          flush(() => {
-            const title = tensorboard.shadowRoot.querySelector(
-              '.toolbar-title'
-            );
-            chai.assert.equal(title.textContent, 'Custom Brand');
-            done();
-          });
+        afterEach(() => {
+          sandbox.restore();
         });
 
-        it('uses customized href for title element ', (done) => {
-          flush(() => {
-            const title = tensorboard.shadowRoot.querySelector(
-              '.toolbar-title'
-            );
-            chai.assert.equal(title.nodeName, 'A');
-            chai.assert.equal(title.href, 'http://custom.home/');
-            done();
-          });
+        it('uses customized brand for title', () => {
+          const title = tensorboard.shadowRoot.querySelector('.toolbar-title');
+          chai.assert.equal(title.textContent, 'Custom Brand');
+        });
+
+        it('uses customized path for title element ', () => {
+          const title = tensorboard.shadowRoot.querySelector('.toolbar-title');
+          chai.assert.equal(title.nodeName, 'A');
+          chai.assert.equal(title.href, 'https://tensorboard.is/awesome');
+        });
+
+        it('does not render the logo as hyperlink for bad protocol', () => {
+          tensorboard.homePath = 'javascript:alert("PWNED!")';
+          chai.assert.isNull(
+            tensorboard.shadowRoot.querySelector('.toolbar-content a')
+          );
```

For consistency with similar tests:

const title = tensorboard.shadowRoot.querySelector('.toolbar-title');
chai.assert.equal(title.nodeName, 'DIV');
chai.assert.isUndefined(title.href);

There are a couple things that worry me about this:
1. Using a different way to assert the same thing across similar tests doubles the chance of test error.
2. Using isNull checks without a similar isNotNull check elsewhere creates risk that your isNull check silently passes for unexpected reasons (for example, the underlying markup changes).

---

I would disagree with your (2). At the end, textContent and pixels are what matters, not the markup. In other words, we are writing BDD (behavior driven development)-style tests and markups do not describe the behavior. They are implementation detail.

---

The code change considerably and no longer uses `assert.isUndefined`.